src/toplevel.c

```diff
@@ -585,8 +592,13 @@ jl_value_t *jl_parse_eval_all(const char *fname, size_t len)
             jl_rethrow();
         }
         else {
-            jl_rethrow_other(jl_new_struct(jl_loaderror_type, fn, ln,
-                                           jl_exception_in_transit));
+            if(jl_typeis(jl_exception_in_transit, jl_initerror_type)) {
```

Comment:
You're right, I think I implicitly assumed this would only get used in one-file-one-module scenarios.

It seems the most appropriate thing would be to run the initializers only after all the code has been loaded, but I suppose that is quite a major change. Is a doubly wrapped exception what makes the most sense then?